src/cargo/ops/cargo_rustc/context/unit_dependencies.rs

```diff
@@ -0,0 +1,301 @@
+//! Constructs the dependency graph for compilation.
+//!
+//! Rust code is typically organized as a set of Cargo packages. The
+//! dependencies between the packages themselves are stored in the
+//! `Resolve` struct. However, we can't use that information as is for
+//! compilation! A package typically contains several targets, or crates,
+//! and these targets has inter-dependencies. For example, you need to
+//! compile the `lib` target before the `bin` one, and you need to compile
+//! `build.rs` before either of those.
+//!
+//! So, we need to lower the `Resolve`, which specifies dependencies between
+//! *packages*, to a graph of dependencies between their *targets*, and this
+//! is exactly what this module is doing! Well, almost exactly: another
+//! complication is that we might want to compile the same target several times
+//! (for example, with and without tests), so we actually build a dependency
+//! graph of `Unit`s, which capture these properties.
+
+use ops::Unit;
+use std::collections::HashMap;
+use CargoResult;
+use core::dependency::Kind as DepKind;
+use ops::{Context, Kind};
+use core::Target;
+use core::Profile;
+
+pub fn build_unit_dependencies<'a, 'cfg>(
+    roots: &[Unit<'a>],
+    cx: &Context<'a, 'cfg>,
+) -> CargoResult<HashMap<Unit<'a>, Vec<Unit<'a>>>> {
+    let mut deps = HashMap::new();
+    for unit in roots.iter() {
+        deps_of(unit, cx, &mut deps)?;
+    }
+
+    Ok(deps)
+}
+
+fn deps_of<'a, 'b, 'cfg>(
+    unit: &Unit<'a>,
+    cx: &Context<'a, 'cfg>,
+    deps: &'b mut HashMap<Unit<'a>, Vec<Unit<'a>>>,
+) -> CargoResult<&'b [Unit<'a>]> {
+    if !deps.contains_key(unit) {
```

Comment:
I may be missing something here, though? Is it required to insert into the map *before* we do the recursive step? I would have figured that the recursion would have been ruled out by this point...

Reply:
`.entry` would not work here, because we need mutable access to `deps` in the call to `calculate_deps` :(

The recursion here is needed only for side effects: we want to fill dependencies not only for root units, but for the whole transitive closure. Ideally, this function should have returned `()`, but we actually call it recursively when we figure out deps for custom_build.

Reply:
Hm ok I'm still not 100% following but that's fine, can clean it up as need be in the future